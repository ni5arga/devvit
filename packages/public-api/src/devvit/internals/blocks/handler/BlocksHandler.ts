import type { UIEvent } from '@devvit/protos';
import { type Effect, type Metadata, type UIRequest, type UIResponse } from '@devvit/protos';
import type { JSONValue } from '@devvit/shared-types/json.js';
import isEqual from 'lodash.isequal';
import type { BlockElement } from '../../../Devvit.js';
import type { ReifiedBlockElement, ReifiedBlockElementOrLiteral } from '../BlocksReconciler.js';
import { BlocksTransformer } from '../BlocksTransformer.js';
import type { EffectEmitter } from '../EffectEmitter.js';
import { ContextBuilder } from './ContextBuilder.js';
import { RenderContext } from './RenderContext.js';
import type { BlocksState, Hook, HookParams, HookSegment, Props } from './types.js';
import { RenderInterruptError } from './types.js';

/**
 * This can be a global/singleton because render is synchronous.
 *
 * If you want to use this from somewhere else, please consider using one of the
 * functions like isRendering or registerHook, and then try to add additional
 * functions here if needed.  Don't use this directly.
 */
export let _activeRenderContext: RenderContext | null = null;

export function useEffectEmitter(): EffectEmitter {
  if (!_activeRenderContext) {
    throw new Error('Hooks can only be declared at the top of a component.');
  }
  return _activeRenderContext;
}

export function isRendering(): boolean {
  return _activeRenderContext !== null;
}

function _structuredClone<T extends JSONValue>(obj: T): T {
  return JSON.parse(JSON.stringify(obj));
}

/**
 * This is the recommended low-level interface for creating hooks like useState or useAsync.
 *
 * Practically, this initializes your hook if it doesn't already exist, and makes sure
 * that its state gets all sync'd up.
 *
 * @param HookSegment -- A name for this hook.  This is used to dedupe hooks.
 * @param initializer
 *    factory for building this hook
 * @returns
 */
export function registerHook<H extends Hook>(
  options: HookSegment,
  initializer: (p: HookParams) => H
): H {
  if (!_activeRenderContext) {
    throw new Error(
      "Hooks can only be declared at the top of a component.  You cannot declare hooks outside\
    of components or inside of event handlers.  It's almost always a mistake to declare hooks inside of loops or\
    conditionals."
    );
  }
  const hookId = _activeRenderContext.nextHookId(options);
  const context = _activeRenderContext;
  const params: HookParams = {
    hookId,
    invalidate: () => {
      context._changed[hookId] = true;
      context._state[hookId] = context?._hooks[hookId]?.state;
    },
    context: _activeRenderContext,
  };
  const fromNull = _activeRenderContext._state[hookId] === undefined;
  _activeRenderContext._hooks[hookId] = _activeRenderContext._hooks[hookId] ?? initializer(params);
  const hook: H = _activeRenderContext._hooks[hookId] as H;

  if (_activeRenderContext._state[hookId] !== undefined) {
    hook.state = _activeRenderContext._state[hookId];
  }
  hook.onStateLoaded?.();
  if (fromNull && hook.state !== undefined && hook.state !== null) {
    params.invalidate();
  }
  return hook;
}

export let _latestBlocksHandler: BlocksHandler | null = null;

/**
 * Replacing BlocksReconciler, the model is now less of a "reconciliation", and more
 * of a handling a request/response lifecycle.
 *
 */
export class BlocksHandler {
  #root: JSX.ComponentFunction;
  #contextBuilder: ContextBuilder = new ContextBuilder();
  #blocksTransformer: BlocksTransformer = new BlocksTransformer(
    () => this._latestRenderContext?.devvitContext?.assets
  );
  _latestRenderContext: RenderContext | null = null;

  constructor(root: JSX.ComponentFunction) {
    if (this.#debug) console.debug('[blocks] BlocksHandler v1');
    this.#root = root;
    _latestBlocksHandler = this;
  }

  async handle(request: UIRequest, metadata: Metadata): Promise<UIResponse> {
    const context = new RenderContext(request, metadata);
    const devvitContext = this.#contextBuilder.buildContext(context, request, metadata);
    context.devvitContext = devvitContext;

    let blocks;

    /**
     * Events on the main queue must be handled in order, so that state is updated in the correct order.  Events
     * on other queues can be handled in parallel, because they only emit effects.
     *
     * There is an optimization here to process SendEventEffects locally, instead of letting them bubble up to the
     * platform.  This prevents a round trip to the platform for every event.
     *
     * This also means we need to respect execution queues here, and not just in the platform.
     */
    const eventsToProcess = request.events;
    const noEvents = !request.events?.length;
    const isMainQueue = noEvents || eventsToProcess.some((e) => !e.async);

    const isBlockingSSR = eventsToProcess.some((e) => e.blocking);

    let changed: { [hookID: string]: true };
    let progress:
      | {
          _state: BlocksState;
          _effects: { [key: string]: Effect };
        }
      | undefined;
    let remaining: UIEvent[] = [...eventsToProcess];

    if (this.#debug) console.debug('[blocks] starting processing events');
    while (eventsToProcess.length > 0) {
      if (this.#debug)
        console.debug('[blocks] processing events loop iteration', eventsToProcess.length);
      /**
       * A concurrently executable batch is a set of events that can be executed in parallel.  This either one main queue event,
       * or any number of other queue events.
       */
      const batch = [];
      if (!eventsToProcess[0].async) {
        batch.push(eventsToProcess.shift()!);
      } else {
        while (eventsToProcess[0]?.async) {
          batch.push(eventsToProcess.shift()!);
        }
      }
      if (!batch.length) throw Error('batch must have at least one event');
      try {
        if (batch[0].async) {
          const stateCopy = _structuredClone(context._state);
          await this.#handleAsyncQueues(context, ...batch);
          // enforce that state updates are only allowed on the main queue.
          context._state = stateCopy;
        } else {
          await this.#handleMainQueue(context, ...batch);
        }
      } catch (e) {
        if (this.#debug) console.debug('[blocks] caught in handler', e);
        /**
         * If we have a progress, we can recover from an error by rolling back to the last progress, and then letting the
         * remaining events be reprocessed.
         */
        if (progress) {
          context._state = progress._state;
          context._changed = changed!;
          context._effects = progress._effects;

          const requeueable = remaining.map((e) => {
            const requeueEvent = { ...e };
            requeueEvent.retry = true;
            return requeueEvent;
          });
          context.addToRequeueEvents(...requeueable);
          break;
        } else {
          throw e;
        }
      }

      if (this.#debug) console.debug('[blocks] remaining events', context._requeueEvents);
      const remainingRequeueEvents: UIEvent[] = [];
      for (const event of context._requeueEvents) {
        if (!isMainQueue && !event.async) {
          if (this.#debug)
            console.debug(
              '[blocks] NOT reprocessing event in BlocksHandler, sync mismatch A',
              event
            );
          // We're async, this is a main qrueue event.  We need to send it back to the platform to let
          // the platform synchronize it.
          remainingRequeueEvents.push(event);
          continue;
        }
        if (isMainQueue && event.async && !isBlockingSSR) {
          if (this.#debug)
            console.debug(
              '[blocks] NOT reprocessing event in BlocksHandler, sync mismatch B',
              event
            );
          // We're main queue, and this is an async event.  We're not in SSR mode, so let's prioritize
          // returning control quickly to the platform so we don't block event loops.
          remainingRequeueEvents.push(event);
          continue;
        }
        if (this.#debug) console.debug('[blocks] reprocessing event in BlocksHandler', event);
        eventsToProcess.push(event);
      }
      context._requeueEvents = remainingRequeueEvents; //

      /**
       * If we're going back through this again, we need to capture the progress, and the remaining events.
       */
      if (eventsToProcess.length > 0) {
        changed = { ...context._changed };
        progress = {
          _state: _structuredClone(context._state),
          _effects: { ...context._effects },
        };
        remaining = [...eventsToProcess];
      }
    } // End of while loop

    if (isMainQueue) {
      const stateCopy = _structuredClone(context._state);
      const eventsCopy = [...context._requeueEvents];
      const effectsCopy = { ...context._effects };

      // Rendering only happens on the main queue.
      const tags = this.#renderRoot(this.#root, context._rootProps ?? {}, context);

      /**
       * It's technically ok for renderRoot to mutate, but that's only in the context of loadHooks.  This render should
       * be idempotent, so we're going to enforce that it doesn't mutate state.
       *
       * TODO: hide this behind a flag, because it's possibly expensive.
       */
      if (!isEqual(context._state, stateCopy)) {
        console.error('[blocks] State was mutated during rendering', context._state, stateCopy);
      }
      if (!isEqual(context._requeueEvents, eventsCopy)) {
        console.error(
          '[blocks] Events were mutated during rendering',
          context._requeueEvents,
          eventsCopy
        );
      }
      if (!isEqual(context._effects, effectsCopy)) {
        console.error(
          '[blocks] Effects were mutated during rendering',
          context._effects,
          effectsCopy
        );
      }

      if (tags) {
        blocks = this.#blocksTransformer.createBlocksElementOrThrow(tags);
        blocks = this.#blocksTransformer.ensureRootBlock(blocks);
      }
    }

    return {
      state: context._changedState,
      effects: context.effects,
      blocks,
      events: context._requeueEvents,
    };
  }

  get #debug(): boolean {
    return !!this._latestRenderContext?.devvitContext.debug.blocks;
  }

  #loadHooks(context: RenderContext, ..._events: UIEvent[]): void {
    // TBD: partial rendering
    context._hooks = {};
    this.#renderRoot(this.#root, context.request.props ?? {}, context);
  }

  /**
   * These can all run in parallel, because they only emit effects
   */
  async #handleAsyncQueues(context: RenderContext, ...batch: UIEvent[]): Promise<void> {
    this.#loadHooks(context, ...batch);

    await Promise.all(
      batch.map(async (event) => {
        if (!event.async) {
          throw new Error(
            "You can't mix main and other queues in one batch.  This is likely a platform bug.  Please file an issue in the Discord for someone to help! https://discord.com/channels/1050224141732687912/1115441897079574620"
          );
        }
        await this.#attemptHook(context, event);
      })
    );
  }

  async #attemptHook(context: RenderContext, event: UIEvent): Promise<void> {
    const hook = context._hooks[event.hook!];
    if (hook?.onUIEvent) {
      try {
        await hook.onUIEvent(event, context);
      } catch (e) {
        console.error('Error in event handler', e);
        throw e;
      }
    } else {
      await context.handleUndeliveredEvent(event);
    }
  }

  async #handleMainQueue(context: RenderContext, ...batch: UIEvent[]): Promise<void> {
    // We need to handle events in order, so that the state is updated in the correct order.
    for (const event of batch) {
      if (this.#debug) console.log('[blocks] handling main queue event', event);
      if (this.#debug) console.log('[blocks] before', context._state);
      this.#loadHooks(context, event);
      await this.#attemptHook(context, event);
      if (this.#debug) console.log('[blocks] after', context._state);
    }

    // TODO: Decide whether this is excessive.  It doesn't hurt anything besides performance.
    this.#loadHooks(context);
  }

  #renderRoot(
    component: JSX.ComponentFunction,
    props: Props,
    context: RenderContext
  ): ReifiedBlockElement | undefined {
    if (this.#debug) console.debug('[blocks] renderRoot');
    context._generated = {};
    _activeRenderContext = context;
    this._latestRenderContext = context;
    try {
      const roots = this.#render(component, props, context);
      if (roots.length !== 1) {
        throw new Error('only one root');
      }
      const root = roots[0];
      if (typeof root === 'string') {
        throw new Error(
          'There must be a root tag.  Try wrapping your app in a <text></text>, <vstack> or other tag.'
        );
      }
      return root;
    } catch (e) {
      if (e instanceof RenderInterruptError) {
        return undefined;
      } else {
        throw e;
      }
    } finally {
      _activeRenderContext = null;
    }
  }

  #render(
    component: JSX.ComponentFunction,
    props: Props,
    context: RenderContext
  ): ReifiedBlockElementOrLiteral[] {
    context.push({ namespace: component.name, ...props });
    try {
      const element = component(props, context.devvitContext);
      return this.#renderElement(element, context);
    } finally {
      context.pop();
    }
  }

  #renderList(list: JSX.Element[], context: RenderContext): ReifiedBlockElementOrLiteral[] {
    list = list.flat(Infinity);
    return list.flatMap((e, i) => {
      if (e && typeof e === 'object' && 'props' in e) {
        if (!e.props?.key) {
          e.props = e.props ?? {};
          e.props.key = `${i}`;
        }
      }
      return this.#renderElement(e, context);
    });
  }

  #renderElement(element: JSX.Element, context: RenderContext): ReifiedBlockElementOrLiteral[] {
    if (Array.isArray(element)) {
      return this.#renderList(element, context);
    } else if (isBlockElement(element)) {
      if (element.type === undefined) {
        return this.#renderList(element.children, context);
      } else if (typeof element.type === 'function') {
        const propsWithChildren = { ...element.props, children: element.children };
        return this.#render(element.type, propsWithChildren, context);
      } else {
        context.push({ namespace: element.type, ...element.props });
        const reifiedChildren = this.#renderList(element.children, context);
        const reifiedProps = this.#reifyProps(element.props ?? {});
        context.pop();
        return [{ type: element.type, children: reifiedChildren, props: reifiedProps }];
      }
    } else {
      return [(element ?? '').toString()];
    }
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  #reifyProps(props: { [key: string]: any }): { [key: string]: JSONValue } {
    const reifiedProps: { [key: string]: JSONValue } = {};
    for (const key in props) {
      if (typeof props[key] === 'undefined') {
        // skip
      } else if (typeof props[key] === 'function') {
        const hook = registerHook(
          {
            namespace: key,
            key: false,
          },
          ({ hookId }) => ({ hookId, state: null, onUIEvent: props[key] })
        );
        reifiedProps[key] = hook.hookId;
        if ('captureHookRef' in props[key]) {
          props[key].captureHookRef();
        }
      } else {
        // push value through the JSON parser to filter incompatible types
        const value = JSON.parse(JSON.stringify(props[key]));
        if (value !== undefined && value !== null) {
          reifiedProps[key] = value;
        }
      }
    }
    return reifiedProps;
  }
}

function isBlockElement(e: JSX.Element): e is BlockElement {
  return typeof e === 'object' && e != null && 'type' in e;
}
