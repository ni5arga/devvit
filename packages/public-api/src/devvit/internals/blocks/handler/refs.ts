import { _activeRenderContext } from './BlocksHandler.js';
import type { HookRef } from './types.js';

/**
 * This function is mostly used for tests.  It's a way to capture the hook id without changing code too much
 *
 * Example:
 *
 * ```tsx
 * const ref: HookRef = {};
 * let [count, setCount] = captureHookRef(useState(0), ref);
 * ```
 *
 * @param value
 * @param ref
 * @returns
 */
export function captureHookRef<T>(value: T, ref: HookRef): T {
  if (!_activeRenderContext) {
    throw new Error('No active render context');
  }

  if (typeof value === 'function') {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    (value as any)['captureHookRef'] = () => {
      if (!_activeRenderContext) {
        throw new Error('No active render context');
      }
      const id = _activeRenderContext._prevHookId;
      ref.id = id;
    };
  } else {
    const id = _activeRenderContext._prevHookId;
    if (!id) {
      throw new Error('No hook id found');
    }
    ref.id = id;
  }

  return value;
}
