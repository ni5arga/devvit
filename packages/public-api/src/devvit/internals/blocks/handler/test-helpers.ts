import type { UIEvent, UIRequest } from '@devvit/protos';
import { Header } from '@devvit/shared-types/Header.js';
import type { JSONValue } from '@devvit/shared-types/json.js';
import { _latestBlocksHandler } from './BlocksHandler.js';
import type { BlocksState, HookRef } from './types.js';

export const findHookId = (ref: HookRef): string => {
  return ref.id!;
};

export const findHookState = (ref: HookRef): JSONValue => {
  return getLatestBlocksState()[ref.id!];
};

export const getLatestBlocksState = (): BlocksState => {
  return _latestBlocksHandler?._latestRenderContext?._state ?? {};
};

export const EmptyRequest: UIRequest = { events: [] };

export const generatePressRequest = (ref: HookRef): UIRequest => {
  const event: UIEvent = {
    hook: ref.id!,
    userAction: {
      actionId: ref.id!,
    },
  };
  return {
    state: getLatestBlocksState(),
    events: [event],
  };
};

export const generateTimerRequest = (ref: HookRef): UIRequest => {
  const event: UIEvent = {
    hook: ref.id!,
    timer: {},
  };
  return {
    state: getLatestBlocksState(),
    events: [event],
  };
};

export const mockMetadata = {
  [Header.AppUser]: {
    values: ['t2_appuser'],
  },
  [Header.Subreddit]: {
    values: ['t5_devvit'],
  },
  [Header.User]: {
    values: ['t2_user'],
  },
};
