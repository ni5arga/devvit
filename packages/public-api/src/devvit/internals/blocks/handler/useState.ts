import type { UIEvent } from '@devvit/protos';
import type { JSONValue } from '@devvit/shared-types/json.js';
import type {
  AsyncUseStateInitializer,
  SetStateAction,
  UseStateInitializer,
  UseStateResult,
} from '../../../../types/hooks.js';
import { registerHook } from './BlocksHandler.js';
import type { RenderContext } from './RenderContext.js';
import type { Hook, HookParams } from './types.js';
import { RenderInterruptError } from './types.js';

export function useState(initialState: UseStateInitializer<boolean>): UseStateResult<boolean>;
export function useState(initialState: UseStateInitializer<number>): UseStateResult<number>;
export function useState(initialState: UseStateInitializer<string>): UseStateResult<string>;
export function useState<S extends JSONValue>(
  initialState: UseStateInitializer<S>
): UseStateResult<S>;
export function useState<S extends JSONValue>(
  initialState: UseStateInitializer<S>
): UseStateResult<S> {
  const hook = registerHook({ namespace: 'useState' }, ({ invalidate: changed }) => {
    const state = initialState instanceof Function ? initialState() : initialState;
    if (state instanceof Promise) {
      throw new Error(
        "Cannot use async initializer with context.useState. Please use useAsyncState instead by importing it: import { Devvit, useAsyncState } from '@devvit/public-api';"
      );
    }
    const setter = (action: SetStateAction<S>): void => {
      hook.state = action instanceof Function ? action(hook.state) : action;
      changed();
    };
    return { state, setter };
  });
  return [hook.state, hook.setter];
}

class AsyncStateHook<S extends JSONValue> implements Hook {
  state: { value: S | null; loading: boolean } = { value: null, loading: false };
  #changed: () => void;
  #ctx: RenderContext;
  #initializer: AsyncUseStateInitializer<S>;
  #hookId: string;

  constructor(initializer: AsyncUseStateInitializer<S>, params: HookParams) {
    this.#initializer = initializer;
    this.#hookId = params.hookId;
    this.#changed = params.invalidate;
    this.#ctx = params.context;
  }

  async onUIEvent(): Promise<void> {
    if (this.state.value === null) {
      this.state.value = await this.#initializer();
      this.state.loading = false;
      this.#changed();
    }
  }

  setter(action: SetStateAction<S>): void {
    this.state.value = action instanceof Function ? action(this.state.value as S) : action;
    this.#changed();
  }

  onStateLoaded(): void {
    if (this.state.value === null) {
      this.state.loading = true;
      this.#changed();
      const requeueEvent: UIEvent = {
        asyncRequest: { requestId: this.#hookId },
        hook: this.#hookId,
      };
      this.#ctx.addToRequeueEvents(requeueEvent);
      throw new RenderInterruptError();
    }
  }
}

export function useAsyncState<S extends JSONValue>(
  initializer: AsyncUseStateInitializer<S>
): UseStateResult<S> {
  const hook = registerHook({ namespace: 'useAsyncState' }, (params) => {
    return new AsyncStateHook(initializer, params);
  });

  return [hook.state.value as S, hook.setter.bind(hook)];
}
