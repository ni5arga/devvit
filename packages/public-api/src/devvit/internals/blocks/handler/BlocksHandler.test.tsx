/** @jsx Devvit.createElement */
/** @jsxFrag Devvit.Fragment */

import { BlockType, type UIRequest } from '@devvit/protos';
import { describe, expect, test, vi } from 'vitest';
import { Devvit } from '../../../Devvit.js';
import { BlocksHandler } from './BlocksHandler.js';
import { captureHookRef } from './refs.js';
import type { HookRef } from './types.js';
import { EmptyRequest, findHookState, generatePressRequest, mockMetadata } from './test-helpers.js';
import { useState } from './useState.js';
import { useAsync } from './useAsync.js';

const funnyRef: HookRef = {};

const loader = async (): Promise<string> => {
  await new Promise((resolve) => setTimeout(resolve, 10));

  return 'hi world!';
};

const yayRef: HookRef = {};
const nayRef: HookRef = {};
const counterRef: HookRef = {};

const SomeAsyncComponent = (): JSX.Element => {
  const { data, error, loading } = captureHookRef(useAsync(loader), funnyRef);
  return (
    <hstack>
      <text>
        {data} - {String(error)} - {loading ? 'loading' : 'done'}
      </text>
    </hstack>
  );
};

const SuccessFail = (): JSX.Element => {
  const [_counter, setCounter] = captureHookRef(useState(0), counterRef);
  return (
    <hstack>
      <button onPress={captureHookRef(() => setCounter((c) => c + 1), yayRef)}>Yay</button>
      <button onPress={captureHookRef(() => Promise.reject('error'), nayRef)}> Nay</button>
    </hstack>
  );
};

describe('BlocksHandler', () => {
  describe('transformation', () => {
    test('should inject a root element', async () => {
      const callback = vi.fn();
      const component = (): JSX.Element => {
        return <button onPress={callback}>Click me</button>;
      };

      const handler: BlocksHandler = new BlocksHandler(component);
      const response = await handler.handle(EmptyRequest, mockMetadata);
      expect(response.blocks?.type).toEqual(BlockType.BLOCK_ROOT);
      expect(response).toMatchSnapshot();
      expect(callback).not.toHaveBeenCalled();
    });
  });
  describe('event shortcuts', () => {
    test('should emit effects in normal operation', async () => {
      const handler = new BlocksHandler(SomeAsyncComponent);
      const response = await handler.handle(EmptyRequest, mockMetadata);
      expect(JSON.stringify(response.blocks)).toContain('loading');

      expect(response.effects.length).toEqual(1);
      expect(response.effects[0].sendEvent?.event?.asyncRequest).toBeDefined();
    });

    test('should process all futures in blocking SSR mode', async () => {
      const handler = new BlocksHandler(SomeAsyncComponent);
      const response = await handler.handle({ events: [{ blocking: {} }] }, mockMetadata);
      expect(JSON.stringify(response.blocks)).toContain('done');
      expect(response.effects.length).toEqual(0);
    });

    test('should be able to batch events and make progress', async () => {
      const handler = new BlocksHandler(SuccessFail);
      await handler.handle(EmptyRequest, mockMetadata);
      expect(findHookState(counterRef)).toEqual(0);
      const req = generatePressRequest(yayRef);
      req.events.push(structuredClone(req.events[0]));
      req.events.push(structuredClone(req.events[0]));
      req.events.push(structuredClone(req.events[0]));
      req.events.push(structuredClone(req.events[0]));
      await handler.handle(req, mockMetadata);
      expect(findHookState(counterRef)).toEqual(5);
    });

    test('should be able to make partial progress', async () => {
      const handler = new BlocksHandler(SuccessFail);
      await handler.handle(EmptyRequest, mockMetadata);
      expect(findHookState(counterRef)).toEqual(0);
      let req = generatePressRequest(nayRef);
      const nay = req.events[0];
      req = generatePressRequest(yayRef);
      req.events.push(structuredClone(req.events[0]));
      req.events.push(structuredClone(req.events[0]));
      req.events.push(nay);
      req.events.push(structuredClone(req.events[0]));

      // yay, yay, yay, NAY, yay
      // 3 successes, one fail, one deferred success
      const response = await handler.handle(req, mockMetadata);
      expect(findHookState(counterRef)).toEqual(3);
      expect(response.effects.length).toEqual(2);
      expect(response.effects[0].sendEvent?.event).toEqual(nay);
      expect(response.effects[1].sendEvent?.event).toEqual(req.events[0]);
      expect(response.effects[1].sendEvent?.jumpsQueue).toEqual(true);
    });

    test('should fail if the first event fails', async () => {
      const handler = new BlocksHandler(SuccessFail);
      await handler.handle(EmptyRequest, mockMetadata);
      expect(findHookState(counterRef)).toEqual(0);
      let req = generatePressRequest(nayRef);
      const nay = req.events[0];
      req = generatePressRequest(yayRef);
      req.events.push(structuredClone(req.events[0]));
      req.events.push(structuredClone(req.events[0]));
      req.events.push(structuredClone(req.events[0]));
      req.events.unshift(nay);

      await expect(handler.handle(req, mockMetadata)).rejects.toThrow('error');
    });
  });

  describe('keys', () => {
    test('should generate nice hook ids', async () => {
      const Box = ({ children }: Devvit.Blocks.HasElementChildren): JSX.Element => {
        const [_counter, _setCounter] = useState(0);
        return <hstack>{children}</hstack>;
      };
      const boxed = (): JSX.Element => (
        <Box>
          <button onPress={() => {}}>hi world</button>
          <button key="kk" onPress={() => {}}>
            hi world
          </button>
          <button id="something" onPress={() => {}}>
            hi world
          </button>
          <button onPress={() => {}}>hi world</button>
        </Box>
      );

      const handler: BlocksHandler = new BlocksHandler(boxed);
      const request = EmptyRequest;
      await handler.handle(request, mockMetadata);
      expect(new Set(Object.keys(handler._latestRenderContext?._generated ?? {}))).toEqual(
        new Set([
          'boxed.Box.useState-0',
          'boxed.Box.hstack.button-0.onPress',
          'boxed.Box.hstack.button-3.onPress',
          'boxed.Box.hstack.button-kk.onPress',
          'something.onPress',
        ])
      );
    });
  });

  describe('function components', () => {
    test('should pass children', async () => {
      const Box = ({ children }: Devvit.Blocks.HasElementChildren): JSX.Element => {
        return <hstack>{children}</hstack>;
      };
      const boxed = (): JSX.Element => (
        <Box>
          <text>hi world</text>
        </Box>
      );
      const expectedComponent = (): JSX.Element => (
        <hstack>
          <text>hi world</text>
        </hstack>
      );
      const handler: BlocksHandler = new BlocksHandler(boxed);
      const request: UIRequest = {
        events: [],
      };
      const response = await handler.handle(request, mockMetadata);
      expect(response).toMatchSnapshot();

      const alt: BlocksHandler = new BlocksHandler(expectedComponent);
      const altResponse = await alt.handle(request, mockMetadata);
      expect(JSON.stringify(altResponse)).toEqual(JSON.stringify(response));
    });
  });

  describe('keys', () => {
    test('keys should be in-sensitive to DOM re-ordering', async () => {
      let which = 0;

      const component: Devvit.BlockComponent = () => {
        const lol = Math.random() > 0.5;
        return (
          <hstack>
            {lol ? (
              <>
                <button
                  key="one"
                  onPress={captureHookRef(() => {
                    which = 1;
                  }, funnyRef)}
                />
                <button key="two" onPress={() => void (which = 2)} />
              </>
            ) : (
              <>
                <button key="two" onPress={() => void (which = 2)} />
                <button
                  key="one"
                  onPress={captureHookRef(() => {
                    which = 1;
                  }, funnyRef)}
                />
              </>
            )}
          </hstack>
        );
      };
      0;
      const handler = new BlocksHandler(component);
      for (let attempts = 0; attempts < 10; attempts++) {
        await handler.handle(EmptyRequest, mockMetadata);
        const req = generatePressRequest(funnyRef);
        await handler.handle(req, mockMetadata);
        expect(Object.keys(handler._latestRenderContext?._generated ?? {}).sort()).toEqual(
          ['component.hstack.button-one.onPress', 'component.hstack.button-two.onPress'].sort()
        );
        expect(which).toBe(1);
      }
    });

    test('ids should be insensitive to re-location', async () => {
      let which = 0;

      const component: Devvit.BlockComponent = () => {
        const lol = Math.random() > 0.5;
        return (
          <hstack>
            {lol ? (
              <hstack>
                <button
                  id="one"
                  onPress={captureHookRef(() => {
                    which = 1;
                  }, funnyRef)}
                />
                <button id="two" onPress={() => void (which = 2)} />
              </hstack>
            ) : (
              <>
                <button id="two" onPress={() => void (which = 2)} />
                <button
                  id="one"
                  onPress={captureHookRef(() => {
                    which = 1;
                  }, funnyRef)}
                />
              </>
            )}
          </hstack>
        );
      };
      const handler = new BlocksHandler(component);

      for (let attempts = 0; attempts < 10; attempts++) {
        await handler.handle(EmptyRequest, mockMetadata);
        const req = generatePressRequest(funnyRef);
        expect(req.events[0].hook).toEqual('one.onPress');
        await handler.handle(req, mockMetadata);
        expect(which).toBe(1);
      }
    });
  });

  describe('actions handler', () => {
    test('should not call the action handler on initial render', async () => {
      const callback = vi.fn();
      const component = (): JSX.Element => {
        return <button onPress={callback}>Click me</button>;
      };

      const handler: BlocksHandler = new BlocksHandler(component);
      const request: UIRequest = {
        events: [],
      };
      const response = await handler.handle(request, mockMetadata);
      expect(response).toMatchSnapshot();
      expect(callback).not.toHaveBeenCalled();
    });

    test('should call the action handler', async () => {
      const callback = vi.fn();
      const component = (): JSX.Element => {
        return <button onPress={callback}>Click me</button>;
      };

      const handler: BlocksHandler = new BlocksHandler(component);
      const request: UIRequest = {
        events: [],
      };
      await handler.handle(request, mockMetadata);
      const hooks = Object.keys(handler._latestRenderContext?._hooks ?? {});
      expect(hooks.length).toBe(1);

      const action = hooks[0];
      expect(callback).not.toHaveBeenCalled();
      expect(action).toMatchSnapshot();

      const actionRequest: UIRequest = {
        events: [
          {
            userAction: {
              actionId: action,
            },
            hook: action,
          },
        ],
      };

      await handler.handle(actionRequest, mockMetadata);
      expect(callback).toHaveBeenCalled();
    });

    test('should not throw when children are or become undefined', async () => {
      const component = (): JSX.Element => {
        return <vstack>{undefined}</vstack>;
      };

      const action = 'non-existent-action';
      const request: UIRequest = {
        events: [
          {
            userAction: {
              actionId: action,
            },
            hook: action,
          },
        ],
      };

      const handler: BlocksHandler = new BlocksHandler(component);
      await handler.handle(request, mockMetadata);
    });
  });
});
