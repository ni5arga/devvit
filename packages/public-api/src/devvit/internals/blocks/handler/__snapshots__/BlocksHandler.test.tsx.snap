// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BlocksHandler > actions handler > should call the action handler 1`] = `"component.button.onPress"`;

exports[`BlocksHandler > actions handler > should not call the action handler on initial render 1`] = `
{
  "blocks": {
    "actions": [],
    "config": {
      "rootConfig": {
        "children": [
          {
            "actions": [
              {
                "data": {},
                "id": "component.button.onPress",
                "type": 0,
              },
            ],
            "config": {
              "buttonConfig": {
                "buttonAppearance": undefined,
                "buttonSize": undefined,
                "disabled": undefined,
                "icon": undefined,
                "text": "Click me",
                "textColor": undefined,
                "textColors": undefined,
              },
            },
            "id": undefined,
            "key": undefined,
            "size": undefined,
            "sizes": undefined,
            "type": 3,
          },
        ],
        "height": 320,
      },
    },
    "id": undefined,
    "key": undefined,
    "size": undefined,
    "sizes": undefined,
    "type": 0,
  },
  "effects": [],
  "events": [],
  "state": {},
}
`;

exports[`BlocksHandler > function components > should pass children 1`] = `
{
  "blocks": {
    "actions": [],
    "config": {
      "rootConfig": {
        "children": [
          {
            "actions": [],
            "config": {
              "stackConfig": {
                "alignment": undefined,
                "backgroundColor": undefined,
                "backgroundColors": undefined,
                "border": undefined,
                "children": [
                  {
                    "actions": [],
                    "config": {
                      "textConfig": {
                        "alignment": undefined,
                        "color": undefined,
                        "colors": undefined,
                        "outline": undefined,
                        "overflow": 1,
                        "selectable": undefined,
                        "size": undefined,
                        "style": undefined,
                        "text": "hi world",
                        "weight": undefined,
                        "wrap": undefined,
                      },
                    },
                    "id": undefined,
                    "key": "0",
                    "size": undefined,
                    "sizes": undefined,
                    "type": 2,
                  },
                ],
                "cornerRadius": undefined,
                "direction": 0,
                "gap": undefined,
                "padding": undefined,
                "reverse": undefined,
              },
            },
            "id": undefined,
            "key": undefined,
            "size": undefined,
            "sizes": undefined,
            "type": 1,
          },
        ],
        "height": 320,
      },
    },
    "id": undefined,
    "key": undefined,
    "size": undefined,
    "sizes": undefined,
    "type": 0,
  },
  "effects": [],
  "events": [],
  "state": {},
}
`;

exports[`BlocksHandler > transformation > should inject a root element 1`] = `
{
  "blocks": {
    "actions": [],
    "config": {
      "rootConfig": {
        "children": [
          {
            "actions": [
              {
                "data": {},
                "id": "component.button.onPress",
                "type": 0,
              },
            ],
            "config": {
              "buttonConfig": {
                "buttonAppearance": undefined,
                "buttonSize": undefined,
                "disabled": undefined,
                "icon": undefined,
                "text": "Click me",
                "textColor": undefined,
                "textColors": undefined,
              },
            },
            "id": undefined,
            "key": undefined,
            "size": undefined,
            "sizes": undefined,
            "type": 3,
          },
        ],
        "height": 320,
      },
    },
    "id": undefined,
    "key": undefined,
    "size": undefined,
    "sizes": undefined,
    "type": 0,
  },
  "effects": [],
  "events": [],
  "state": {},
}
`;
